### PR TITLE
Ferigis.24.gen fsm as a workers

### PR DESCRIPTION
**Pull request by ferigis**
This is my fix for #24 
If you agree with this approach, the README file should be edited too, in order to specify the relative fsm stuff.
Please, provide feedback.

### CODE REVIEW

File: src/wpool_fsm_process.erl [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+age(Process) -> gen_fsm:sync_send_all_state_event(Process, age).
+
+%%%===================================================================
+%%% init, terminate, code_change, info callbacks
+%%%===================================================================
+%% @private
+-spec init({atom(), atom(), term(), [wpool:option()]}) ->
+  {ok, dispatch_state, state()}.
+init({Name, Mod, InitArgs, Options}) ->
+  case Mod:init(InitArgs) of
```
Comment (gadgetci) [inline src/wpool_fsm_process.erl]: According to **Elvis**:
> Remove the dynamic function call on line 91. Only modules that define callbacks should make dynamic calls.

File: src/wpool_fsm_process.erl [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+    Error -> Error
+  end.
+
+%% @private
+-spec terminate(atom(), fsm_state(), state()) -> term().
+terminate(Reason,
+    CurrentState,
+    #state{mod=Mod, state=Mod_State, name=Name, options=Options}) ->
+  ok = notify_queue_manager(worker_dead, Name, Options),
+  Mod:terminate(Reason, CurrentState, Mod_State).
```
Comment (gadgetci) [inline src/wpool_fsm_process.erl]: According to **Elvis**:
> Remove the dynamic function call on line 116. Only modules that define callbacks should make dynamic calls.

File: src/wpool.erl [hunk 3]
```diff
@@ -48,6 +48,10 @@
 -export([start_pool/1, start_pool/2, start_sup_pool/1, start_sup_pool/2]).
 -export([stop_pool/1]).
 -export([call/2, cast/2, call/3, cast/3, call/4, call/5]).
+-export([send_event/2, send_event/3, sync_send_event/2, sync_send_event/3,
+  sync_send_event/4, sync_send_event/5, send_all_event/2, send_all_event/3,
+  sync_send_all_event/2, sync_send_all_event/3, sync_send_all_event/4,
```
Comment (elbrujohalcon) [inline src/wpool.erl]: Why do you rename some functions and not others?
I would either go the whole way and rename all the functions like [`ktn_fsm`](https://github.com/inaka/erlang-katana/blob/master/src/ktn_fsm.erl) does or keep the original names, like `sync_send_all_state_event`, for instance.

File: src/wpool.erl [hunk 4]
```diff
@@ -156,6 +160,108 @@ cast(Sup, Cast, {hash_worker, HashKey}) ->
...
+%% @equiv sync_send_event(Sup, Event, default_strategy())
+-spec sync_send_event(name(), term()) -> term().
+sync_send_event(Sup, Event) -> sync_send_event(Sup, Event, default_strategy()).
+
+%% @equiv sync_send_event(Sup, Event, Strategy, 5000)
+-spec sync_send_event(name(), term(), strategy()) -> term().
+sync_send_event(Sup, Event, Strategy) ->
+  sync_send_event(Sup, Event, Strategy, 5000).
+
+%% @doc Picks a server and issues the event to it.
```
Comment (elbrujohalcon) [inline src/wpool.erl]: Picks an *fsm*

File: src/wpool.erl [hunk 5]
```diff
@@ -156,6 +160,108 @@ cast(Sup, Cast, {hash_worker, HashKey}) ->
...
+-spec sync_send_all_event(name(), term()) -> term().
+sync_send_all_event(Sup, Event) ->
+  sync_send_all_event(Sup, Event, default_strategy()).
+
+%% @equiv sync_send_all_event(Sup, Event, Strategy, 5000)
+-spec sync_send_all_event(name(), term(), strategy()) -> term().
+sync_send_all_event(Sup, Event, Strategy) ->
+  sync_send_all_event(Sup, Event, Strategy, 5000).
+
+%% @doc Picks a server and issues the event to it.
```
Comment (elbrujohalcon) [inline src/wpool.erl]: Picks an *fsm*

File: src/wpool_fsm_process.erl (hunk 2 above)
Comment (gadgetci) [inline src/wpool_fsm_process.erl]: According to **Elvis**:
> Remove the dynamic function call on line 116. Only modules that define callbacks should make dynamic calls.

**Comment by elbrujohalcon** [comment]
@ferigis I like this approach. I added a few comments, but in general it looks good.
I'm not sure why @gadgetci is complaining, tho… @jfacorro will check that out
I assigned the PR to @igaray as well for another round of code review.

File: src/wpool.erl [hunk 6]
```diff
@@ -156,6 +161,121 @@ cast(Sup, Cast, {hash_worker, HashKey}) ->
...
+sync_send_all_state_event(Sup, Event, Strategy, Timeout) ->
+  wpool_fsm_process:sync_send_all_state_event(wpool_pool:Strategy(Sup),
+    Event, Timeout).
+
+%% @doc Picks a server and issues the event to it.
+%%      For all strategies except available_worker, WorkerTimeout is ignored
+%%      For available_worker the total timeout will be WorkerTimeout + Timeout.
+-spec sync_send_all_state_event(name(), term(), strategy(),
+                          available_worker_timeout(), timeout()) -> term().
+sync_send_all_state_event(Sup
```
Comment (gadgetci) [inline src/wpool.erl]: According to **Xref**:
> `wpool:sync_send_all_state_event/5` calls undefined function `wpool_pool:sync_send_all_state_event_to_available_worker/3`

File: src/wpool.erl [hunk 7]
```diff
@@ -156,6 +161,121 @@ cast(Sup, Cast, {hash_worker, HashKey}) ->
...
+%% @doc Picks a fsm and issues the event to it.
+%%      For all strategies except available_worker, Timeout applies only to the
+%%      time spent on the actual call to the worker, because time spent finding
+%%      the worker in other strategies is negligible.
+%%      For available_worker the time used choosing a worker is also considered
+-spec sync_send_all_state_event(name()
+                                , term()
+                                , strategy()
+                                , timeout()) -> term().
+sync_send_all_state_event(Sup, Event, available_worker, Timeout) ->
```
Comment (gadgetci) [inline src/wpool.erl]: According to **Xref**:
> `wpool:sync_send_all_state_event/4` calls undefined function `wpool_pool:sync_send_all_state_event_to_available_worker/3`

File: src/wpool.erl [hunk 8]
```diff
@@ -156,6 +161,121 @@ cast(Sup, Cast, {hash_worker, HashKey}) ->
...
+  wpool_fsm_process:send_event(wpool_pool:Strategy(Sup), Event).
+
+%% @equiv send_all_state_event(Sup, Event, default_strategy())
+-spec send_all_state_event(name(), term()) -> ok.
+send_all_state_event(Sup, Event) ->
+  send_all_state_event(Sup, Event, default_strategy()).
+
+%% @doc Picks a server and issues the event to it
+-spec send_all_state_event(name(), term(), strategy()) -> ok.
+send_all_state_event(Sup, Event, available_worker) ->
```
Comment (gadgetci) [inline src/wpool.erl]: According to **Xref**:
> `wpool:send_all_state_event/3` calls undefined function `wpool_pool:send_all_state_event_to_available_worker/2`

File: src/wpool_fsm_process.erl [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
```
Comment (gadgetci) [inline src/wpool_fsm_process.erl]: According to **Elvis**:
> Remove the dynamic function call on line 121. Only modules that define callbacks should make dynamic calls.

File: src/wpool_fsm_process.erl [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+age(Process) -> gen_fsm:sync_send_all_state_event(Process, age).
+
+%%%===================================================================
+%%% init, terminate, code_change, info callbacks
+%%%===================================================================
+%% @private
+-spec init({atom(), atom(), term(), [wpool:option()]}) ->
+  {ok, dispatch_state, state()}.
+init({Name, Mod, InitArgs, Options}) ->
+  case Mod:init(InitArgs) of
```
Comment (gadgetci) [inline src/wpool_fsm_process.erl]: According to **Elvis**:
> Remove the dynamic function call on line 96. Only modules that define callbacks should make dynamic calls.

**Comment by ferigis** [comment]
@elbrujohalcon , I have just committed my changes regarding your suggestions. Thanks

**Comment by elbrujohalcon** [comment]
@igaray it's all yours

**Pull request merged**